kedro/runner/runner.py

```diff
@@ -242,54 +241,86 @@ def _find_persistent_ancestors(
         ``Node``s.
 
     """
-    ancestor_nodes_to_run = set()
+    initial_nodes_to_run: set[Node] = set()
+
     queue, visited = deque(children), set(children)
     while queue:
         current_node = queue.popleft()
-        if _has_persistent_inputs(current_node, catalog):
-            ancestor_nodes_to_run.add(current_node)
+        impersistent_inputs = _enumerate_impersistent_inputs(current_node, catalog)
+
+        # If all inputs are persistent, we can run this node as is
+        if not impersistent_inputs:
+            initial_nodes_to_run.add(current_node)
             continue
-        for parent in _enumerate_parents(pipeline, current_node):
-            if parent in visited:
+
+        # Otherwise, look for the nodes that produce impersistent inputs
+        for node in _enumerate_nodes_with_outputs(pipeline, impersistent_inputs):
+            if node in visited:
                 continue
-            visited.add(parent)
-            queue.append(parent)
-    return ancestor_nodes_to_run
+            visited.add(node)
+            queue.append(node)
+
+    return initial_nodes_to_run
+
+
+def _enumerate_impersistent_inputs(node: Node, catalog: DataCatalog) -> set[str]:
+    """Enumerate impersistent input Datasets of a ``Node``.
+
+    Args:
+        node: the ``Node`` to check the inputs of.
+        catalog: the ``DataCatalog`` of the run.
 
+    Returns:
+        Set of names of impersistent inputs of given ``Node``.
+
+    """
+    # We use _data_sets because they pertain parameter name format
+    catalog_datasets = catalog._datasets
+    missing_inputs: set[str] = set()
+    for node_input in node.inputs:
+        # Important difference vs. Kedro approach
+        if node_input.startswith("params:"):
+            continue
+        if isinstance(catalog_datasets[node_input], MemoryDataset):
```

Comment:
Can we use `catalog.load` instead of `catalog._datasets`? If i understand you want to perserve `params:` and `catalog.load` should work

Reply:
Is it necessary to load the dataset though? Isn't that a bit of overkill?

Reply:
In this approach I need to check the Dataset instance, not exactly its data contents